### PR TITLE
Add Set-Cookie fingerprint for Laravel

### PR DESCRIPTION
**Pull request by adfoster-r7**
## Description

Adds a new `Set-Cookie` fingerprint for Laravel - a PHP web application framework

## How Has This Been Tested?

Run the server form vulhub:

```
$ docker run -p 9001:80 vulhub/laravel:8.4.2
```

Curl output example:

```
$ curl -I http://localhost:9001
HTTP/1.1 200 OK
Date: Mon, 13 Mar 2023 16:59:08 GMT
Server: Apache/2.4.38 (Debian)
X-Powered-By: PHP/7.4.15
Cache-Control: no-cache, private
Set-Cookie: XSRF-TOKEN=eyJpdiI6Im1KOUxDMC9BeTJCZkllSklkYXF3eGc9PSIsInZhbHVlIjoibDFoQ1JZcVdobjdyaHQ2SjNoR1dQaS94Qnk5THIxYzNKclZGekVEVzd0VHE0T3NBNkN1Y0FqaGxsVVg0SzhGV3NKM1dNSzlkc2NabDBHT2twcUVFMTNmL3JMMGxLSDh1cUlaQzhqMEZpckpDbkx0REg2UkNtR21rRkw5N05WR1ciLCJtYWMiOiJkZTZhMTc2ZTMzZmMwODYyOGE1OWRkN2U0Y2JjOTI5MDZiY2UzNGZhNjQ4OGU5ZDY3NDkzZjQ3NjYwYmViOTVkIn0%3D; expires=Mon, 13-Mar-2023 18:59:08 GMT; Max-Age=7200; path=/; samesite=lax
Set-Cookie: laravel_session=eyJpdiI6ImtMa0k3VHNIVXBkMlZlc28xV3pFNnc9PSIsInZhbHVlIjoiWUhPTkFjSGZlWmVUNyt1OU9qd0t1SDdSZmhVNDVVbFBMWXBKVnhDN1ozYW5aZFNObTdLQ0tQNEFWMC90OU9CaDBPaU00SklBZ2hmRGNZREwxcVBMRlQ1ay85bmx6NHBkcEt2MWdIR1JSQjVpNFNKS3QzY1ZpcGdSYWt5L1FQcnAiLCJtYWMiOiI1MmM0ODZmYWE1MWY5NDkwZDVmN2NmMmZlZjA4MjNjMjljZDRjYzkxY2FmMjI5NTY3YTA4YjYxMzc4ZDJiYWFmIn0%3D; expires=Mon, 13-Mar-2023 18:59:08 GMT; Max-Age=7200; path=/; httponly; samesite=lax
Content-Type: text/html; charset=UTF-8
```

Running through recog:

```
$ curl --silent -I http://localhost:9001 | grep -i '^Set-Cookie:' | cut -d: -f2- | bin/recog_match xml/http_cookies.xml - | grep MATCH
MATCH: {"matched"=>"Laravel PHP web application framework", "service.vendor"=>"laravel", "service.product"=>"laravel", "service.cpe23"=>"cpe:/a:laravel:laravel:-", "service.protocol"=>"http", "fingerprint_db"=>"http_header.cookie", "data"=>"laravel_session=eyJpdiI6IkZhSGdNakRVUUtpS1ZYck5wcTl1cGc9PSIsInZhbHVlIjoiR3FlWVFXTXIzNW5BRVhoRHhkMk5NYnhzcVovSU4vL1JZRXVDbEpWUnM5TWs0MW0zWmRORTAzVFZjcVljd2lidHVsWWZ3aGNvRXdMTFNBcThadFRMKytyU20vV3p0T0xaTEI3NnpNU1Z5WW9ySkxQaDBMMkFSRDNJK3J0QmtWL3kiLCJtYWMiOiIwZmNlOTIwZjkxMThhZGUzZDViZDcwNjJkMzE4MWEzMDE1OWJjNTAwNTZiYjBkODMwZDliYTlhOWFiZjAyMTU5In0%3D; expires=Mon, 13-Mar-2023 18:59:21 GMT; Max-Age=7200; path=/; httponly; samesite=lax"}
```

Note Laravel doesn't have favicon by default, it's just an empty file, so I didn't include that fingerprint - https://github.com/laravel/laravel/blob/9ae75b58a1ffc00ad36bf1e877fe2bf9ec601b82/public/favicon.ico

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have updated the documentation accordingly (or changes are not required).
- [x] I have added tests to cover my changes (or new tests are not required).
- [x] All new and existing tests passed.

### CODE REVIEW

File: xml/http_cookies.xml [hunk 1]
```diff
@@ -681,6 +681,14 @@
     <param pos="0" name="service.certainty" value="0.5"/>
   </fingerprint>
 
+  <fingerprint pattern="^laravel_session=">
+    <description>Laravel PHP web application framework</description>
+    <example>laravel_session=eyJpdiI6IjZmcXM4SXZqNVQxY2U2UDVEb1V0SFE9PSIsInZhbHVlIjoiQlRUYXk2WEsxUkZmVU5lRXpBTlZkUURkUnphR0RGMTdaRDkxWHFLVGhtaUZwM0plaWR2aVMyeVQ5MmFzNlVFczVQUExpeDdhUUZveVlCOVUxMVA4dHp6ZUQrK0t3KzBUaVlxb2o4WUxleE1rb29GTlozZWF6cXpUUTRnNFpJTHUiLCJtYWMiOiI3ZGNhNzhmOGRmMWFmZmE5ZjFiYjQ5ODBlNGEwMjgzZWNiZGI2ZTI4ZjFmYWJkMmIxZTgwYzM2NjA1OThlNGNlIn0%3D; expires=Mon, 13-Mar-2023 18:21:13 GMT; Max-Age=7200; path=/; httponly; samesite=lax</example>
+    <param pos="0" name="service.vendor" value="laravel"/>
+    <param pos="0" name="service.product" value="laravel"/>
```
Comment (adfoster-r7) [inline xml/http_cookies.xml:688]: I grabbed this from:

> Version 2.3: cpe:2.3:a:laravel:laravel:8.4.2:*:*:*:*:*:*:*
> https://nvd.nist.gov/products/cpe/detail/62651740-3F6C-410D-9C06-9F71B0D077E5?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Alaravel&status=FINAL%2CDEPRECATED

Interestingly I also found there is CPE for `framework` as the product, but I'm not sure why both co-exist:

> Version 2.3: cpe:2.3:a:laravel:framework:8.4.0:*:*:*:*:*:*:*
> https://nvd.nist.gov/products/cpe/detail/92B54D53-6C45-4877-BBD6-9885AB6FC19F?namingFormat=2.3&orderBy=CPEURI&keyword=cpe%3A2.3%3Aa%3Alaravel&status=FINAL%2CDEPRECATED

Comment (mkienow-r7) [reply xml/http_cookies.xml:688]: Also, verify that you are not looking at deprecated entries.

File: xml/http_cookies.xml [hunk 2]
```diff
@@ -834,4 +842,4 @@
     <param pos="0" name="service.certainty" value="0.0"/>
   </fingerprint>
 
-</fingerprints>
\ No newline at end of file
+</fingerprints>
```
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply xml/http_cookies.xml:845]: Will revert this thanks :+1:

Out of interest, is there a a historical reason for this? If I understand correctly the posix spec [3.206](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206) requires a terminating newline character for files, and XML parsers should handle the extra newline gracefully

For context - I had run recog_cleanup on the file which doesn't seem to to strip the new line https://github.com/rapid7/recog/blob/0975a6c1840ccd163ff0beb3832677fb5a314831/bin/recog_cleanup#L8-L11

I could create a PR to add the trailing new lines and update the recog_cleanup tool to enforce the newline?

File: xml/http_cookies.xml [hunk 3]
```diff
@@ -681,6 +681,14 @@
     <param pos="0" name="service.certainty" value="0.5"/>
   </fingerprint>
 
+  <fingerprint pattern="^laravel_session=">
+    <description>Laravel PHP web application framework</description>
+    <example>laravel_session=eyJpdiI6IjZmcXM4SXZqNVQxY2U2UDVEb1V0SFE9PSIsInZhbHVlIjoiQlRUYXk2WEsxUkZmVU5lRXpBTlZkUURkUnphR0RGMTdaRDkxWHFLVGhtaUZwM0plaWR2aVMyeVQ5MmFzNlVFczVQUExpeDdhUUZveVlCOVUxMVA4dHp6ZUQrK0t3KzBUaVlxb2o4WUxleE1rb29GTlozZWF6cXpUUTRnNFpJTHUiLCJtYWMiOiI3ZGNhNzhmOGRmMWFmZmE5ZjFiYjQ5ODBlNGEwMjgzZWNiZGI2ZTI4ZjFmYWJkMmIxZTgwYzM2NjA1OThlNGNlIn0%3D; expires=Mon, 13-Mar-2023 18:21:13 GMT; Max-Age=7200; path=/; httponly; samesite=lax</example>
+    <param pos="0" name="service.vendor" value="laravel"/>
```
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply xml/http_cookies.xml:687]: I originally aligned with the value in the CPE - but will update :+1:

File: xml/http_cookies.xml (hunk 1 above)
Comment (adfoster-r7) [reply xml/http_cookies.xml:688]: Thanks! I've looked at this again, and I believe this is not deprecated, and is the one I want 🤞

File: identifiers/service_product.txt [hunk 4]
```diff
@@ -283,6 +283,7 @@ LDAP Agent for eDirectory
 LDAP Server
 LLBServer
 Lansweeper
+Laravel
```
Comment (adfoster-r7) [inline identifiers/service_product.txt:286]: Generated with `./bin/recog_standardize --write xml/*.xml`

File: xml/http_cookies.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkienow-r7) [reply xml/http_cookies.xml]: Historical in the sense of following from the PR guidance of previous maintainers. `bin/recog_cleanup` does correct new line at the end of the file. For example, I introduced a new line at the end of xml/http_wwwauth.xml and ran cleanup. Please disregard changes cleanup will introduce to xml/telnet_banners.xml. I have a work in progress rewrite of the cleanup script I need to get back to and finish.

```
$ git diff
diff --git a/xml/http_wwwauth.xml b/xml/http_wwwauth.xml
index be75820..dd4fe97 100644
--- a/xml/http_wwwauth.xml
+++ b/xml/http_wwwauth.xml
@@ -870,4 +870,5 @@ Digest realm="WatchGuard SOHO  Configuration",qop="auth",nonce="7c8e98007db66888
      Basic realm="User Login"
   -->

-</fingerprints>
\ No newline at end of file
+</fingerprints>
+
$ ruby bin/recog_cleanup
$ git diff xml/http_wwwauth.xml
```

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply xml/http_cookies.xml]: Ah, I believe there was a race condition between running `bin/recog_cleanup` and my editor's auto-save functionality re-adding the line behind the scenes. Patched now, thanks :+1:

**Review by mkienow-r7 — APPROVED** [review]
Thank you for the contribution @adfoster-r7!

**Pull request merged**